### PR TITLE
Numerous parsing fixes

### LINKED ISSUE #314: Backticks and escaping
Backslash escaping should be suppressed **inside** code span. Backtick, which would otherwise be an opener of a code span, still may be escaped.

Input:
```
\`x`
```

Expected output:
```
<p>`x`</p>
```

Actual output:
```
<p><code>x</code></p>
```

### PR DESCRIPTION
**Pull request by marcusklaas**
Fixes #314, #315 and #317.

Many thanks to @mity for reporting these issues with test cases. Such work makes fixing these things much easier! :bowing_man:

### CODE REVIEW

**Comment by marcusklaas** [comment]
Also fixes https://github.com/raphlinus/pulldown-cmark/issues/294 now.

**Comment by marcusklaas** [comment]
Also closes #296 and #304.

File: src/parse.rs [hunk 1]
```diff
@@ -651,10 +647,11 @@ impl<'a> FirstPass<'a> {
                 b'`' => {
                     self.tree.append_text(begin_text, ix);
                     let count = 1 + scan_ch_repeat(&bytes[ix+1..], b'`');
+                    let preceeded_by_backslash = ix > 0 && bytes[ix - 1] == b'\\';
                     self.tree.append(Item {
                         start: ix,
                         end: ix + count,
-                        body: ItemBody::MaybeCode(count),
+                        body: ItemBody::MaybeCode(count, preceeded_by_backslash),
```
Comment (marcusklaas) [inline src/parse.rs]: Hmmmm, this is probably not correct. `` \\`foo` `` may now incorrectly render to ``<p>\`foo`</p>`` where it should create a proper span in this case.

Comment (mity) [reply src/parse.rs]: Also be aware of funny case as this one:
```
\``foo``bar`
```

Note the 1st backtick is escaped, but the 2nd still makes beginning of a code span - of length 1 in this case so the closer is after the `bar`, not before it.

**Comment by marcusklaas** [comment]
Yup, it's very tricky. I *think* we managed such situations ok, but after seeing what other (trivial) tests we failed, I'm not sure about anything any more ;-)

Thanks again for your valuable input. I will check your example as soon as I am back on my machine.

File: src/parse.rs (hunk 1 above)
Comment (marcusklaas) [reply src/parse.rs]: It produces `<p><code>foo``bar</code></p>` now. So the code span is alright, but there should be a text node with a single backtick preceding it, right?

**Comment by marcusklaas** [comment]
` ``` ` renders to `<code></code>` in both pulldown-cmark as the reference implementation. That makes no sense to me :thinking:

**Comment by marcusklaas** [comment]
Ok, the backtick issues should be fixed now. Also https://github.com/raphlinus/pulldown-cmark/issues/300.

**Comment by mity** [comment]
> `` ``` `` renders to `<code></code>` in both pulldown-cmark as the reference implementation. That makes no sense to me 🤔

It's required by the specs. See https://spec.commonmark.org/0.29/#example-98. 

If the enclosing block ends, it enforces the end of the unfinished fenced code block. Here the enclosing block is the document as a whole.

**Comment by marcusklaas** [comment]
Oooh wait those are code *blocks*, not inline code. It makes sense now! :sweat_smile: 

Thanks for the reference!

File: src/parse.rs (hunk 1 above)
Comment (mity) [reply src/parse.rs]: Yes.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/293 now.

**Comment by marcusklaas** [comment]
Addresses https://github.com/raphlinus/pulldown-cmark/issues/320.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/318.

**Comment by marcusklaas** [comment]
Addresses https://github.com/raphlinus/pulldown-cmark/issues/283.

**Comment by marcusklaas** [comment]
Hmmm, performance does seem to slip a bit with these changes (2-3%).. Will have to dig a bit to see what is causing this..

**Comment by marcusklaas** [comment]
Closes https://github.com/raphlinus/pulldown-cmark/issues/307.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/328.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/303 and #313.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/319.

**Comment by marcusklaas** [comment]
Addresses https://github.com/raphlinus/pulldown-cmark/issues/289.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/330.

**Comment by marcusklaas** [comment]
Closes https://github.com/raphlinus/pulldown-cmark/issues/292.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/331.

**Comment by marcusklaas** [comment]
Closes https://github.com/raphlinus/pulldown-cmark/issues/326.

**Comment by marcusklaas** [comment]
Fixes https://github.com/raphlinus/pulldown-cmark/issues/327.

**Comment by marcusklaas** [comment]
Addresses https://github.com/raphlinus/pulldown-cmark/issues/329.

**Comment by marcusklaas** [comment]
Closes https://github.com/raphlinus/pulldown-cmark/issues/295.

File: src/parse.rs [hunk 2]
```diff
@@ -1482,18 +1486,24 @@ impl InlineStack {
                 min(mod3_lower, self.lower_bounds[1])
             }
         } else {
-            0
+            self.lower_bounds[5]
```
Comment (raphlinus) [inline src/parse.rs]: Might be a good idea to define constants for these magic numbers, now that there are a few of them.

Comment (marcusklaas) [reply src/parse.rs]: :+1: done!

**Comment by marcusklaas** [comment]
Thanks for the review. And especially deep mega thanks to @oberien and @mity for opening so many hiqh quality issues with reproducible test cases! That really helped a lot! :bowing_man:

**Pull request merged**